README.md
# luno-api-node

A simple wrapper for the Luno API. The module supports promise and callbacks.

> This module is forked from https://github.com/bausmeier/node-bitx. 
> Since the original repository is not maintained anymore, starting with version v2.0.0 this fork has been detached from original repository. It is now stand alone. Contributions are not pull-requested to original repository.

## Usage
Add luno as a dependency:

```bash
$ npm install --save luno-api-node
```


### Luno([keyId, keySecret][, options])
To access the private Luno API methods you must supply your key id and key secret as the first two arguments. If you are only accessing the public API endpoints you can leave these two arguments out.

The optional `options` argument can be used to override the default options. The default options are equivalent to:

```js
{
  hostname: 'api.luno.com',
  port: 443,
  ca: undefined,
  pair: 'XBTZAR'
}
```

## Properties

### apiCallRate

Gives the current rate per minute of API calls. The property is read only.
It can be used by applications to limit the API call rate to prevent HTTP error code `429` (too many requests) responses from Luno server.
See Luno API documentation for applicable rate limitations: https://www.luno.com/en/developers/api#tag/Rate-Limiting

> API calls which result in error code `429` or `ErrTooManyRequests` are considered failed/unprocessed server request and are not added to the `apiCallRate` counter

Example:

```javascript
console.log(luno.apiCallRate)
```


## Methods
For details about the API endpoints see https://www.luno.com/en/developers/api.

### Callbacks
The arguments passed to the callback function for each method are:

1. An error or `null` if no error occurred.
2. An object containing the data returned by the Luno API.

The `callback` function is optional. If the `callback` is not provided, the methods return a promise. 

### createAccount(currency, name[, callback])
POST https://api.luno.com/api/1/accounts

Example:

```javascript
luno.createAccount('XBT', 'Trading ACC', function (err, response) {})
```


### updateAccountName(currency, name[, callback])
PUT https://api.luno.com/api/1/accounts/{id}/name

Example:

```javascript
luno.updateAccountName(12345, 'Trading ACC', function (err, response) {})
```


### listAccountPendingTransactions(id[, callback])
GET https://api.luno.com/api/1/accounts/{id}/pending

Example:

```javascript
luno.listAccountPendingTransactions(12345, function (err, response) {})
```


### listAccountTransactions(id, min_row, max_row[, callback])
GET https://api.luno.com/api/1/accounts/{id}/transactions

Example:

```javascript
luno.listAccountTransactions(12345, 1, 1000, function (err, response) {})
```


### listAccountBalances([assets][, callback])
GET https://api.luno.com/api/1/balance

Example:

```javascript
luno.listAccountBalances(['XBT', 'ETH'], function (err, response) {})
```

### getMove(options[, callback])
GET https://api.luno.com/api/exchange/1/move

Example:

```javascript
luno.getMove({ id: 18563829047 }, function (err, response) {})
```

### move(amount, debit_account_id, credit_account_id [,options] [, callback])
POST https://api.luno.com/api/exchange/1/move

Example:

```javascript
luno.move('10000', 12345, 12346, { client_move_id: 'mv-53960812' }, function (err, response) {})
```

### listMoves([options][, callback])
GET https://api.luno.com/api/exchange/1/move/list_moves

Example:

```javascript
luno.listMoves({ before: 1530865703508, limit: 986 }, function (err, response) {})
```


### getTicker([options][, callback])
GET https://api.luno.com/api/1/ticker/XBTZAR

Default options:

```javascript
{
  pair: luno.pair
}
```

Example:

```javascript
luno.getTicker(function (err, ticker) {})
```

### getAllTickers([callback])
GET https://api.luno.com/api/1/tickers

Example:

```javascript
luno.getAllTickers(function (err, tickers) {})
```

### getOrderBook([options][, callback])
GET https://api.luno.com/api/1/orderbook

Default options:

```javascript
{
  pair: luno.pair
}
```

Example:

```javascript
luno.getOrderBook(function (err, orderBook) {})
```

### getTrades([options][, callback])
GET https://api.luno.com/api/1/trades

Default options:

```javascript
{
  pair: luno.pair
}
```

Example:

```javascript
luno.getTrades(function (err, trades) {})
```

### getTradeList([options][, callback])
GET https://api.luno.com/api/1/listtrades

Default options:

```javascript
{
  pair: luno.pair
}
```

Example:

```javascript
luno.getTradeList({ sort_desc: true, limit: 10 }, function (err, tradeList) {})
```

### getOrderList([options][, callback])
GET https://api.luno.com/api/1/listorders

Default options:

```javascript
{
  pair: luno.pair,
  state: undefined
}
```

Example:

```javascript
luno.getOrderList({ state: 'PENDING' }, function (err, orderList) {})
```

### getOrderListV2([options][, callback])
GET https://api.luno.com/api/exchange/2/listorders

Default options:

```javascript
{
  pair: luno.pair,
  state: undefined
}
```

Example:

```javascript
luno.getOrderListV2({ closed: true }, function (err, orderList) {})
```

### getOrderListV3(options[, callback])
GET https://api.luno.com/api/exchange/3/order

Example:

```javascript
luno.getOrderListV3({ id: 'BXMC2CJ7HNB88U4' }, function (err, orderList) {})
```

### getBalance([asset][, callback])
GET https://api.luno.com/api/1/balance

Example:

```javascript
luno.getBalance('ZAR', function (err, balance) {})
```

### getFundingAddress(asset[, options][, callback])
GET https://api.luno.com/api/1/funding_address

Default options:

```javascript
{
  address: undefined
}
```

Example:

```javascript
luno.getFundingAddress('XBT', { address: 'B1tC0InExAMPL3fundIN6AdDreS5t0Use' }, function (err, fundingAddress) {})
```

### createFundingAddress(asset[, callback])
POST https://api.luno.com/api/1/funding_address

Example:

```javascript
luno.createFundingAddress('XBT', function (err, fundingAddress) {})
```

### getFeeInfo([options][, callback])
GET https://api.luno.com/api/1/fee_info

Default options:

```javascript
{
  pair: luno.pair
}
```

Example:

```javascript
luno.getFeeInfo({ pair: 'XBTZAR' }, function (err, feeInfo) {})
```

### postBuyOrder(volume, price[, options][, callback])
POST https://api.luno.com/api/1/postorder

Example:

```javascript
luno.postBuyOrder(9999.99, 0.01, function (err, order) {})
```

### postSellOrder(volume, price[, options][, callback])
POST https://api.luno.com/api/1/postorder

Example:

```javascript
luno.postSellOrder(0.01, 9999.99, function (err, order) {})
```

### postMarketBuyOrder(volume[, options][, callback])
POST https://api.luno.com/api/1/marketorder

Example:

```javascript
luno.postMarketBuyOrder(0.01, function (err, order) {})
```

### postMarketSellOrder(volume[, options][, callback])
POST https://api.luno.com/api/1/marketorder

Example:

```javascript
luno.postMarketSellOrder(0.01, function (err, order) {})
```

### stopOrder(orderId[, callback])
POST https://api.luno.com/api/1/stoporder

Example:

```javascript
luno.stopOrder('BXMC2CJ7HNB88U4', function (err, result) {})
```

### getOrder(orderId[, callback])
GET https://api.luno.com/api/1/orders/{orderId}

Example:

```javascript
luno.getOrder('BXHW6PFRRXKFSB4', function (err, result) {})
```

### getOrderV2(orderId[, callback])
GET https://api.luno.com/api/exchange/2/orders/{orderId}

Example:

```javascript
luno.getOrderV2('BXHW6PFRRXKFSB4', function (err, result) {})
```

### getOrderV3(options[, callback])
GET https://api.luno.com/api/exchange/3/orders/{orderId}

Example:

```javascript
luno.getOrderV3({ id: 'BXMC2CJ7HNB88U4' }, function (err, result) {})
luno.getOrderV3({ client_order_id: 'lmt-53960812' }, function (err, result) {})
```

### getTransactions(asset[, options][, callback])
GET https://api.luno.com/api/1/transactions

Default options:
```javascript
{
  offset: 0,
  limit: 10
}
```

Example:

```javascript
luno.getTransactions('XBT', { offset: 5, limit: 20 }, function (err, transactions) {})
```

### getWithdrawals([callback])
GET https://api.luno.com/api/1/withdrawals

Example:

```javascript
luno.getWithdrawals(function (err, withdrawals) {})
```

### getWithdrawal(withdrawalId[, callback])
GET https://api.luno.com/api/1/withdrawals/{withdrawalId}

Example:

```javascript
luno.getWithdrawal('1212', function (err, withdrawal) {})
```

### requestWithdrawal(type, amount[, callback])
POST https://api.luno.com/api/1/withdrawals

Example:

```javascript
luno.requestWithdrawal('ZAR_EFT', 1000, function (err, withdrawal) {})
```

### cancelWithdrawal(withdrawalId[, callback])
DELETE https://api.luno.com/api/1/withdrawals/{withdrawalId}

Example:

```javascript
luno.cancelWithdrawal('1212', function (err, withdrawal) {})
```

### listTransfers(id[, options][, callback])
GET https://api.luno.com/api/exchange/1/transfers

Example:

```javascript
luno.listTransfers(1212, { limit: 986 }, function (err, data) {})
```

## Contributing

Open a pull request or create an issue and help me improve it.


CHANGELOG.md
### 2.2.0 (2022-12-21)

* add method `listTransfers()`
* fix url path for `getMove()`, `move()`, `listMoves()` and `updateAccountName()`
* update documentation

### 2.1.0 (2022-12-02)

* add sub-account methods: `createAccount()`, `updateAccountName()`, `listAccountPendingTransactions()`, `listAccountTransactions()`, `listAccountBalances()`, `getMove()`, `move()`, `listMoves()`
* fix method `getBalance()` to take optional parameter `asset` as string or array
* remove method `getTransaction()`, which was deprecated
* update tests

### 2.0.1 (2022-07-14)

* Add debug option to output API call requests and responses 

### 2.0.0 (2022-07-14)

* Repository detached from original (and the fork is now stand alone)
* all Bitx usages renamed to Luno
* Add `apiCallRate` value to the error message, when an `ErrTooManyRequests` occurs
* remove method `getLimits()`

### 1.10.0 (2022-06-20)

* Add `apiCallRate` property
* Update tests

### 1.9.1 (2022-06-20)

* Fix https request error handling when no `callback` parameter (promises)

### 1.9.0 (2021-12-06)

* Add methods `getOrderListV2()`, `getOrderV2()`, `getOrderV3()`

### 1.8.0 (2020-10-13)

* Add promise support for all methods. When `callback` parameter is not provided, the methods return a promise

### 1.7.1

* Parse Luno API errors
* Update domain to `api.luno.com`

### 1.7.0

* Add getTradeList method

### 1.6.0

* Add options to `postBuyOrder()`, `postMarketBuyOrder()`, `postSellOrder()` and `postMarketSellOrder()`
* Add `getFeeInfo()` method

### 1.5.1

* Fix an issue with GET requests sent after POST requests

### 1.5.0

* Add `getOrder()` method

### 1.4.2

* Make `asset` parameter optional for `getBalance()`

### 1.4.1

* Only catch JSON.parse errors (#6)

### 1.4.0

* Add `pair` option to `getTicker()`, `getOrderBook()` and `getTrades()`
* Add `getAllTickers()` method

### 1.3.0

* Add `createFundingAddress()` method
* Add `getTransactions()` method
* Accept `address` option for `getFundingAddress()`
* Accept `state` option for `getOrderList()`
* Add `getWithdrawals()` method
* Add `getWithdrawal()` method
* Add `requestWithdrawal()` method
* Add `cancelWithdrawal()` method

### 1.2.1

* Fix path bug (#1)

### 1.2.0

* Add `getFundingAddress()` method

### 1.1.0

* Support multiple currency pairs
* `BitX.getLimits()` is now deprecated (use `BitX.getBalance()` instead).
* Updated to the latest BitX API.
* Added `BitX.getBalance()` method.

### 1.0.1

* Set response encoding to utf8
* Resolve require path properly


test/UnitTests.js
'use strict'

const Luno = require('../lib/luno')
const FakeRequest = require('./helpers/FakeRequest')
const https = require('https')
const sinon = require('sinon')
const tap = require('tap')

tap.test('Luno constructor', {autoend: true}, function (t) {
  t.test('should create a new instance', function (tt) {
    const luno = new Luno()
    tt.type(luno, 'Luno')
    tt.end()
  })

  tap.test('should create a new instance without the new keyword', function (tt) {
    const luno = Luno()
    tt.type(luno, 'Luno')
    tt.end()
  })

  tap.test('should have default options', function (tt) {
    const luno = new Luno()
    tt.equal(luno.hostname, 'api.luno.com')
    tt.equal(luno.port, 443)
    tt.equal(luno.pair, 'XBTZAR')
    tt.end()
  })

  tap.test('should accept options', function (tt) {
    const options = {
      hostname: 'localhost',
      port: 8000,
      pair: 'XBTUSD'
    }
    const luno = new Luno(options)
    tt.equal(luno.hostname, options.hostname)
    tt.equal(luno.port, options.port)
    tt.equal(luno.pair, options.pair)
    tt.end()
  })

  tap.test('should accept auth and options', function (tt) {
    const keyId = 'cnz2yjswbv3jd'
    const keySecret = '0hydMZDb9HRR3Qq-iqALwZtXLkbLR4fWxtDZvkB9h4I'
    const options = {
      hostname: 'localhost',
      port: 8000,
      pair: 'XBTUSD'
    }
    const luno = new Luno(keyId, keySecret, options)
    tt.equal(luno.hostname, options.hostname)
    tt.equal(luno.port, options.port)
    tt.equal(luno.pair, options.pair)
    tt.equal(luno.auth, keyId + ':' + keySecret)
    tt.end()
  })
})

tap.test('Internal', {autoend: true}, function (t) {
  const keyId = '12345'
  const keySecret = '0000000000000000'
  const path = '/api/1/test'

  const expectedOptions = {
    headers: {
      'Accept': 'application/json',
      'Accept-Charset': 'utf-8'
    },
    hostname: 'api.luno.com',
    path: path,
    port: 443,
    auth: keyId + ':' + keySecret
  }

  const luno = new Luno(keyId, keySecret)
  const request = sinon.stub(https, 'request')
  const stub = request.withArgs(sinon.match(expectedOptions))

  t.afterEach(function () {
    stub.reset()
  })

  t.teardown(function () {
    request.restore()
  })

  t.test('_request should return the expected result', function (tt) {
    const expectedResult = {success: true}
    stub.returns(new FakeRequest(expectedResult))
    luno._request('GET', path, null, function (err, result) {
      tt.error(err)
      tt.same(result, expectedResult)
      tt.end()
    })
  })

  t.test('_request should return an error if request emits an error', function (tt) {
    stub.returns(new FakeRequest(null, {fail: true}))
    luno._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })

  t.test('_request should return an error if the response is not valid', function (tt) {
    stub.returns(new FakeRequest('invalid', {stringify: false}))
    luno._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })

  t.test('_request should return an error if the response contains an error', function (tt) {
    stub.returns(new FakeRequest({error: true}))
    luno._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })

  t.test('_request should return an error if the response is unauthorized', function (tt) {
    stub.returns(new FakeRequest(null, {statusCode: 401}))
    luno._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })
})

tap.test('External', {autoend: true}, function (t) {
  const luno = new Luno()
  const callback = function () {}
  let mock

  t.beforeEach(function () {
    mock = sinon.mock(luno)
  })

  t.afterEach(function () {
    mock.restore()
  })

  t.test('getTicker should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/ticker', {pair: 'XBTZAR'}, callback)
    luno.getTicker(callback)
    mock.verify()
    tt.end()
  })

  t.test('getTicker should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/ticker', {pair: 'XBTMYR'}, callback)
    luno.getTicker({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('getAllTickers should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/tickers', null, callback)
    luno.getAllTickers(callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderBook should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/orderbook', {pair: 'XBTZAR'}, callback)
    luno.getOrderBook(callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderBook should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/orderbook', {pair: 'XBTMYR'}, callback)
    luno.getOrderBook({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('getTrades should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/trades', {pair: 'XBTZAR'}, callback)
    luno.getTrades(callback)
    mock.verify()
    tt.end()
  })

  t.test('getTrades should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/trades', {pair: 'XBTMYR'}, callback)
    luno.getTrades({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderList should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/listorders', {pair: 'XBTZAR'}, callback)
    luno.getOrderList(callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderList should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/listorders', {pair: 'XBTMYR'}, callback)
    luno.getOrderList({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('postBuyOrder should call _request with the correct parameters', function (tt) {
    const parameters = {
      type: 'BID',
      volume: 9999.99,
      price: 0.001,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/postorder', parameters, callback)
    luno.postBuyOrder(parameters.volume, parameters.price, callback)
    mock.verify()
    tt.end()
  })

  t.test('postMarketBuyOrder should call _request with the correct parameters', function (tt) {
    const parameters = {
      type: 'BUY',
      counter_volume: 9999.99,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/marketorder', parameters, callback)
    luno.postMarketBuyOrder(parameters.counter_volume, callback)
    mock.verify()
    tt.end()
  })

  t.test('postSellOrder should call _request with the correct parameters', function (tt) {
    const parameters = {
      type: 'ASK',
      volume: 0.001,
      price: 9999.99,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/postorder', parameters, callback)
    luno.postSellOrder(parameters.volume, parameters.price, callback)
    mock.verify()
    tt.end()
  })

  t.test('postMarketSellOrder should call _request with the correct parameters', function (tt) {
    const parameters = {
      type: 'SELL',
      base_volume: 9999.99,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/marketorder', parameters, callback)
    luno.postMarketSellOrder(parameters.base_volume, callback)
    mock.verify()
    tt.end()
  })

  t.test('stopOrder should call _request with the correct parameters', function (tt) {
    const body = {order_id: 'BXMC2CJ7HNB88U4'}
    mock.expects('_request').once().withArgs('POST', '/api/1/stoporder', body, callback)
    luno.stopOrder(body.order_id, callback)
    mock.verify()
    tt.end()
  })

  t.test('getBalance should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/balance', null, callback)
    luno.getBalance(callback)
    mock.verify()
    tt.end()
  })

  t.test('getBalance should accept an asset argument and call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/balance', { assets: ['ZAR'] }, callback)
    luno.getBalance('ZAR', callback)
    mock.verify()
    tt.end()
  })

  t.test('should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/funding_address', { asset: 'XBT' }, callback)
    luno.getFundingAddress('XBT', callback)
    mock.verify()
    tt.end()
  })
})


test/Regression.js
'use strict'

const Luno = require('../lib/luno')
const https = require('https')
const fs = require('fs')
const path = require('path')
const tap = require('tap')

tap.test('GET after POST', (t) => {
  const options = {
    key: fs.readFileSync(path.join(__dirname, 'ssl', 'server', 'test.key')),
    cert: fs.readFileSync(path.join(__dirname, 'ssl', 'server', 'test.crt'))
  }
  const server = https.createServer(options).listen(process.env.PORT || 0, 'localhost', function () {
    const address = server.address()
    const luno = new Luno('keyId', 'keySecret', {
      hostname: 'localhost',
      port: address.port,
      ca: fs.readFileSync(path.join(__dirname, 'ssl', 'ca', 'root.pem'))
    })

    t.afterEach(function () {
      server.removeAllListeners('request')
    })

    t.teardown(function (done) {
      server.close(done)
    })

    t.test('should not send content type or content length headers', function (t) {
      server.once('request', function (req, res) {
        req.resume()
        req.on('end', function () {
          res.end(JSON.stringify({}))
        })
      })

      luno.postBuyOrder(9999.99, 0.0001, function (err, order) {
        t.error(err)

        server.once('request', function (req, res) {
          t.notOk(req.headers['content-type'])
          t.notOk(req.headers['content-length'])
          res.end(JSON.stringify({}))
        })

        luno.getTicker(function (err, ticker) {
          t.error(err)
          t.end()
        })
      })
    })

    t.end()
  })
})


lib/BitX.js
'use strict'

const https = require('https')
const path = require('path')
const querystring = require('querystring')
const config = require(path.join(__dirname, '..', 'package'))

const defaultHeaders = {
  'Accept': 'application/json',
  'Accept-Charset': 'utf-8',
  'User-Agent': 'node-bitx v' + config.version
}

function BitX (keyId, keySecret, options) {
  if (!(this instanceof BitX)) {
    return new BitX(keyId, keySecret, options)
  }

  if (typeof keyId === 'string') {
    this.auth = keyId + ':' + keySecret
  } else {
    options = keyId
  }

  options = options || {}
  this.hostname = options.hostname || 'api.luno.com'
  this.port = options.port || 443
  this.ca = options.ca
  this.pair = options.pair || 'XBTZAR'
  this._requestMts = []
  this._requestsRefresh = this._requestsRefresh.bind(this)
  this._timeoutId = null
}

Object.defineProperty(BitX.prototype, 'apiCallRate', {
  get: function () {
    return this._requestsRefresh()
  }
})

BitX.prototype._requestsRefresh = function () {
  const maxMts = Date.now() - 60000
  while (this._requestMts.length > 0 && this._requestMts[0] < maxMts) {
    this._requestMts.shift()
  }

  if (this._requestMts.length > 0) {
    this._timeoutId.refresh()
    this._timeoutId.unref()
  } else {
    clearTimeout(this._timeoutId)
    this._timeoutId = null
  }

  return this._requestMts.length
}

BitX.prototype._request = function (method, resourcePath, data, callback) {
  const headers = Object.assign({}, defaultHeaders)
  data = querystring.stringify(data)
  if (method === 'GET') {
    if (data) {
      resourcePath += '?' + data
    }
  } else if (method === 'POST') {
    headers['Content-Type'] = 'application/x-www-form-urlencoded'
    headers['Content-Length'] = Buffer.byteLength(data)
  }

  const options = {
    headers: headers,
    hostname: this.hostname,
    path: resourcePath,
    port: this.port,
    auth: this.auth,
    method: method,
  }

  if (this.ca) {
    options.ca = this.ca
    options.agent = new https.Agent(options)
  }

  const mts = Date.now()
  const requestMts = this._requestMts
  requestMts.push(mts)
  if (this._timeoutId === null) {
    this._timeoutId = setTimeout(this._requestsRefresh, 60000)
    this._timeoutId.unref()
  }

  const apiCallRate = this.apiCallRate
  const promise = new Promise(function (resolve, reject) {
    const req = https.request(options)
    req.on('response', function (res) {
      let response = ''

      res.setEncoding('utf8')

      res.on('data', function (data) {
        response += data
      })

      res.on('end', function () {
        if (res.statusCode !== 200) {
          let lunoApiError
          try {
            response = JSON.parse(response)
            let errTooManyRequestsDetails = ''
            if (response.error_code.includes('429') || response.error_code.includes('ErrTooManyRequests')) {
              let i = requestMts.length - 1
              while (i >= 0) {
                if (requestMts[i] === mts) {
                  requestMts.splice(i, 1)
                  break
                }

                i -= 1
              }

              errTooManyRequestsDetails = ' (' + apiCallRate + ')'
            }

            lunoApiError = new Error('luno API error ' + response.error_code +': ' + errTooManyRequestsDetails + response.error)
            lunoApiError.error_code = response.error_code
          } catch (err) {
            lunoApiError = new Error('luno API error ' + res.statusCode + ': ' + response)
          }

          if (lunoApiError) {
            if (typeof callback === 'function') {
              return callback(lunoApiError)
            } else {
              return reject(lunoApiError)
            }
          }
        }

        try {
          response = JSON.parse(response)
        } catch (err) {
          if (typeof callback === 'function') {
            return callback(err)
          } else {
            return reject(err)
          }
        }

        if (response.error) {
          const err = new Error(response.error)
          if (typeof callback === 'function') {
            return callback(err)
          } else {
            return reject(err)
          }
        }

        if (typeof callback === 'function') {
          return callback(null, response)
        } else {
          return resolve(response)
        }

      })
    })

    req.on('error', function (err) {
      if (typeof callback === 'function') {
        return callback(err)
      } else {
        return reject(err)
      }
    })

    if (method === 'POST' && data) {
      req.write(data)
    }

    req.end()
  })

  if (typeof callback !== 'function') {
    return promise
  }
}

BitX.prototype.getTicker = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    pair: this.pair,
  }

  return this._request('GET', '/api/1/ticker', Object.assign(defaults, options), callback)
}

BitX.prototype.getAllTickers = function (callback) {
  return this._request('GET', '/api/1/tickers', null, callback)
}

BitX.prototype.getOrderBook = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    pair: this.pair,
  }

  return this._request('GET', '/api/1/orderbook', Object.assign(defaults, options), callback)
}

BitX.prototype.getTrades = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    pair: this.pair,
  }

  return this._request('GET', '/api/1/trades', Object.assign(defaults, options), callback)
}

BitX.prototype.getOrderList = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    pair: this.pair,
  }

  return this._request('GET', '/api/1/listorders', Object.assign(defaults, options), callback)
}

BitX.prototype.getOrderListV2 = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    pair: this.pair,
  }

  return this._request('GET', '/api/exchange/2/listorders', Object.assign(defaults, options), callback)
}

BitX.prototype.getTradeList = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    pair: this.pair,
  }

  return this._request('GET', '/api/1/listtrades', Object.assign(defaults, options), callback)
}

BitX.prototype.getLimits = function (callback) {
  console.log('node-bitx warning: BitX.getLimits is deprecated. Please use BitX.getBalance instead.')
  return this._request('GET', '/api/1/BTCZAR/getlimits', null, callback)
}

BitX.prototype.getFeeInfo = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    pair: this.pair,
  }

  return this._request('GET', '/api/1/fee_info', Object.assign(defaults, options), callback)
}

BitX.prototype.stopOrder = function (orderId, callback) {
  const body = {
    order_id: orderId,
  }

  return this._request('POST', '/api/1/stoporder', body, callback)
}

BitX.prototype.postBuyOrder = function (volume, price, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const body = {
    type: 'BID',
    volume: volume,
    price: price,
    pair: this.pair,
  }

  return this._request('POST', '/api/1/postorder', Object.assign(body, options), callback)
}

BitX.prototype.postMarketBuyOrder = function (volume, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const body = {
    type: 'BUY',
    counter_volume: volume,
    pair: this.pair,
  }

  return this._request('POST', '/api/1/marketorder', Object.assign(body, options), callback)
}

BitX.prototype.postSellOrder = function (volume, price, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const body = {
    type: 'ASK',
    volume: volume,
    price: price,
    pair: this.pair,
  }

  return this._request('POST', '/api/1/postorder', Object.assign(body, options), callback)
}

BitX.prototype.postMarketSellOrder = function (volume, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const body = {
    type: 'SELL',
    base_volume: volume,
    pair: this.pair,
  }

  return this._request('POST', '/api/1/marketorder', Object.assign(body, options), callback)
}

BitX.prototype.getOrder = function (id, callback) {
  return this._request('GET', '/api/1/orders/' + id, null, callback)
}

BitX.prototype.getOrderV2 = function (id, callback) {
  return this._request('GET', '/api/exchange/2/orders/' + id, null, callback)
}

BitX.prototype.getOrderV3 = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  return this._request('GET', '/api/exchange/3/order', options, callback)
}

BitX.prototype.getBalance = function (asset, callback) {
  if (typeof asset === 'function') {
    callback = asset
    asset = null
  }

  return this._request('GET', '/api/1/balance', asset ? {asset: asset} : null, callback)
}

BitX.prototype.getFundingAddress = function (asset, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    asset: asset,
  }

  return this._request('GET', '/api/1/funding_address', Object.assign(defaults, options), callback)
}

BitX.prototype.createFundingAddress = function (asset, callback) {
  return this._request('POST', '/api/1/funding_address', {asset: asset}, callback)
}

BitX.prototype.getTransactions = function (asset, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }

  const defaults = {
    asset: asset,
    offset: 0,
    limit: 10,
  }

  return this._request('GET', '/api/1/transactions', Object.assign(defaults, options), callback)
}

BitX.prototype.getWithdrawals = function (callback) {
  return this._request('GET', '/api/1/withdrawals/', null, callback)
}

BitX.prototype.getWithdrawal = function (id, callback) {
  return this._request('GET', '/api/1/withdrawals/' + id, null, callback)
}

BitX.prototype.requestWithdrawal = function (type, amount, callback) {
  const options = {
    type: type,
    amount: amount,
  }

  return this._request('POST', '/api/1/withdrawals/', options, callback)
}

BitX.prototype.cancelWithdrawal = function (id, callback) {
  return this._request('DELETE', '/api/1/withdrawals/' + id, null, callback)
}

module.exports = BitX
